<a href="https://colab.research.google.com/github/dscohen75/DS-Unit-1-Sprint-2-Statistics/blob/master/module4/LS_DS_124_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 124

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):` 

You should only truly need these three values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?


In [4]:
# TODO - code!

def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate=0.08, true_positive_rate=1):
  
  non_drunk_prior = 1 - prob_drunk_prior

  posterior = true_positive_rate * prob_drunk_prior / (true_positive_rate * prob_drunk_prior + false_positive_rate * non_drunk_prior)

  print(posterior)


In [5]:
prob_drunk_given_positive(.001)

0.012357884330202669


## The probability that a person is drunk after one breathalizer test is 0.012.


In [6]:
prob_drunk_given_positive(0.012357884330202669)

0.13525210993291495


## After two positive breathalizer tests, the probability that the person is truly drunk is 0.135.


In [7]:
prob_drunk_given_positive(0.13525210993291495)

0.6615996951348605


In [8]:
prob_drunk_given_positive(0.6615996951348605)

0.9606895076105054


## It would take four positive breathilizer tests in a row in order to say that the probability that this person is drunk is greater than 95%.

### 2) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).

The frequentist approach uses a sample of data from a known (or assumed) distribution. We use a likelihood distribution, which tells us the *probability of getting this sample of data, assuming a hypothesis.* For a given sample, a hypothesis test will always turn out the same way and depends upon sample size.

The Bayesian approach is to use prior assumptions about *the probability our hypothesis is true*, and then update that probability to incorporate new information brought in by our data. 

### 3) Use the following Template to help come up with ideas for your Build Sprint Project: 



---

## Idea 1:

### You
What do you care about?

What do you know about?

What decisions do you face?




What I care about: Lifestyle habits that impact long-term quality of life

What I know about: Human physiology

What decisions I face: What habits to focus on in my life in order to improve my sleep quality

### Seven templates

In order to better understand the prompts here, please review the [Priceonomics Content Marketing Templates](https://priceonomics.com/introducing-priceonomics-content-marketing/)

Can you apply the templates to your topics?

**Geographic Variation:**

**Trend related to the news:**

**Who does that?:**

**Answering a question people care about:**

**Valuable to businesses:**

**What's the most popular?:**

**Cost/Money rankings:**

How does amount of sleep vary across different people? Who engages in various habits/ behaviors that may be related to sleep quality?

What physiological markers are associated with sleep quantity and quality?

How are the rates of certain diseases related to sleep quality and quantity?

### Misconceptions

What misconceptions do people have about your topic?

People are not aware of just how important sleep is for all aspects of health and wellbeing. There is actually a stigma or "machismo" surrounding sleep, that it's a waste of time or unnecessary, leading to the "I'll sleep when I'm dead" approach.

Even those who are aware of the importance and want to get better sleep may not realize or believe that they can do anything to impact their sleep quality.

### Examples

What data storytelling example inspires you?

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

https://www.nature.com/articles/s41539-019-0055-z

https://www.nature.com/articles/s41562-019-0754-8

These two articles tell compelling stories. One of them says that the data are available by contacting the corresponding authors.

###Data

Where could you search for data about your topic?

## Kaggle, Awesome Data Sets, ??

---

## Idea 2:

### You
What do you care about?

What do you know about?

What decisions do you face?



## What I care about: Life impacts of chronic pain

## What I know about: The beneifts of physical therapy for chronic pain as an alternative to opioid medications or surgery.

## What decisions I face: (In the past) How to help people with chronic pelvic pain to improve their quality of life.

### Seven templates

In order to better understand the prompts here, please review the [Priceonomics Content Marketing Templates](https://priceonomics.com/introducing-priceonomics-content-marketing/)

Can you apply the templates to your topics?

**Geographic Variation:**

**Trend related to the news:**

**Who does that?:**

**Answering a question people care about:**

**Valuable to businesses:**

**What's the most popular?:**

**Cost/Money rankings:**

## What are the most common treatment approaches for pelvic pain, by geographical region?
## What are the sources that people turn to for help most often when suffering with pelvic pain?
## What is the cost of each of these approaches?
## How much do each of these treatment approaches improve peoples' quality of life?


### Misconceptions

What misconceptions do people have about your topic?

## 1. That chronic pain is unsolvable and will never improve.
## 2. That the results of an MRI or other scan will diagnose the source of their pain.
## 3. That medications and surgery are the most effective or only options for treating chronic pain.

### Examples

What data storytelling example inspires you?

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

https://painisreallystrange.wordpress.com/2015/03/16/spinal-degeneration-not-linked-to-pain-on-big-imaging-review/

## There are many examples of the above kind of story, showing that back pain or neck pain is not correlated with MRI findings, and therefore scans may not be the best first step to address these problems.

###Data

Where could you search for data about your topic?

## Maybe I could contact the authors of one of these studies and ask whether they'd be willing to share their data with me for this project.

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 & 3 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!